gnovm/pkg/gnolang/type_check.go

```diff
@@ -215,6 +215,38 @@ func assertAssignableTo(xt, dt Type, autoNative bool) {
 	}
 }
 
+// checkValConstType checks the type and values are valid for a ValueDecl of const.
+func checkValConstType(d *ValueDecl) {
+	if d.Type != nil {
+		switch d.Type.(type) {
+		case *BasicLitExpr, *NameExpr:
+			// Valid constant type expression, in case of NameExpr should evaluate if underlying type is a basic type
+		default:
+			panic("const type should be a basic type")
+		}
+	}
+	for _, vx := range d.Values {
+		checkValConstValue(vx)
+	}
+}
+
+// checkValConstValue validate an expr is valid as a value for a const
+func checkValConstValue(expr Expr) {
+	switch x := expr.(type) {
+	case *BasicLitExpr, *NameExpr:
+		// Valid constant type expression, in case of NameExpr should evaluate if underlying type is a basic type
+	case *BinaryExpr:
+		checkValConstValue(x.Left)
```

Comment:
also this one:
```go
package main


var x = 1

const ff = 1 << x

func main() {
	println("ok")
}
```

Reply:
Hello @ltzmaxwell,

I understand better thank you, however not all RHS are constExpr by default and this code is then executed to transform them at the last moment.

```go
if n.Const {
	// NOTE: may or may not be a *ConstExpr,
	// but if not, make one now.
	for i, vx := range n.Values {
		n.Values[i] = evalConst(store, last, vx)
	}
}
```

So i tried to just change this code by adding a `if !isConst(vx) -> panic` but some expr are not yet evaluted as const.

I'm wondering whether I should try to evaluate all constant expressions as `constExpr` before reaching the snippet of code above or continue along the path of checking each type of expression, which can be complicated by the fact that special cases exist in many cases?

I also noticed that this code evaluates the const `Minute` as a `constExpr`.

```
package main

import (
	"fmt"
)

type Duration int64

const (
	Nanosecond  Duration = 1
	Microsecond          = 1000 * Nanosecond
	Millisecond          = 1000 * Microsecond
	Second               = 1000 * Millisecond
	Minute               = 60 * Second
	Hour                 = 60 * Minute
)

const df = Minute

func main() {
	fmt.Printf("df: %v %T\n", df, df)
}
```

but this code where i import the pkg time does not evalue the RHS `time.Duration` of the const decl. as a constExpr before the final loop, which transforms the RHS into constExpr regardless of its value.

```
package main

import (
	"fmt"
	"time"
)

const df = time.Minute

func main() {
	fmt.Printf("df: %v %T\n", df, df)
}
```

It's one of my problems right now.
I closed the PR for now since i think i need more vision and knowledge about the whole architecture to fix the problem, i would be happy to peer-code the fix or see a proposal for this issue.